builder/etcd/etcd.go

```diff
@@ -0,0 +1,524 @@
+		if err != nil || res.Node == nil {
+			continue
+		}
+
+		content := res.Node.Value
+		if err := ioutil.WriteFile(path, []byte(content), 0600); err != nil {
+			log.Errf(c, "Error writing ssh host key file: %s", err)
+		}
+	}
+
+	// Now get generic key.
+	res, err := client.Get("sshHostKey", false, false)
+	if err != nil || res.Node == nil {
+		return fmt.Errorf("Failed to get sshHostKey from etcd. %v", err)
+	}
+
+	content := res.Node.Value
+	if err := ioutil.WriteFile("/etc/ssh/ssh_host_key", []byte(content), 0600); err != nil {
+		log.Errf(c, "Error writing ssh host key file: %s", err)
+		return err
+	}
+	return nil
+}
+
+// keysToEtcd copies local keys into etcd.
+//
+// It only fails if it cannot copy ssh_host_key to sshHostKey. All other
+// abnormal conditions are logged, but not considered to be failures.
+func keysToEtcd(c cookoo.Context, client Setter, ciphers []string, etcdPath string) error {
+	lpath := "/etc/ssh/ssh_host_%s_key"
+	privkey := "%s/sshHost%sKey"
+	for _, cipher := range ciphers {
+		path := fmt.Sprintf(lpath, cipher)
+		key := fmt.Sprintf(privkey, etcdPath, cipher)
+		content, err := ioutil.ReadFile(path)
+		if err != nil {
+			log.Infof(c, "No key named %s", path)
+		} else if _, err := client.Set(key, string(content), 0); err != nil {
+			log.Errf(c, "Could not store ssh key in etcd: %s", err)
+		}
+	}
+	// Now we set the generic key:
+	if content, err := ioutil.ReadFile("/etc/ssh/ssh_host_key"); err != nil {
+		log.Errf(c, "Could not read the ssh_host_key file.")
+		return err
+	} else if _, err := client.Set("sshHostKey", string(content), 0); err != nil {
+		log.Errf(c, "Failed to set sshHostKey in etcd.")
+		return err
+	}
+	return nil
+}
+
+// genSshKeys generates the default set of SSH host keys.
+func genSSHKeys(c cookoo.Context) {
+	// Generate a new key
+	out, err := exec.Command("ssh-keygen", "-A").CombinedOutput()
+	if err != nil {
+		log.Infof(c, "ssh-keygen: %s", out)
+		log.Errf(c, "Failed to generate SSH keys: %s", err)
+	}
+}
+
+// UpdateHostPort intermittently notifies etcd of the builder's address.
+//
+// If `port` is specified, this will notify etcd at 10 second intervals that
+// the builder is listening at $HOST:$PORT, setting the TTL to 20 seconds.
+//
+// This will notify etcd as long as the local sshd is running.
+//
+// Params:
+// 	- base (string): The base path to write the data: $base/host and $base/port.
+// 	- host (string): The hostname
+// 	- port (string): The port
+// 	- client (Setter): The client to use to write the data to etcd.
+// 	- sshPid (int): The PID for SSHD. If SSHD dies, this stops notifying.
+func UpdateHostPort(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+
+	base := p.Get("base", "").(string)
+	host := p.Get("host", "").(string)
+	port := p.Get("port", "").(string)
+	client := p.Get("client", nil).(Setter)
+	sshd := p.Get("sshdPid", 0).(int)
+
+	// If no port is specified, we don't do anything.
+	if len(port) == 0 {
```

Comment:
Good point.

```
package a

import "testing"

func TestX(t *testing.T) {
}

func BenchmarkEqual(b *testing.B) {
	str := ""
	for i := 0; i < b.N; i++ {
		if str == "" {
			// do something
		}
	}
}

func BenchmarkLen(b *testing.B) {
	str := ""
	for i := 0; i < b.N; i++ {
		if len(str) == 0 {
			// do something
		}
	}
}
```

```
kalbasit@wks-113 ~/code/src/github.com/icza/gowut/gwu [master] ± % go test -v -bench=. -run=XXX /tmp/a_test.go
PASS
BenchmarkEqual-8	2000000000	         0.34 ns/op
BenchmarkLen-8  	2000000000	         0.32 ns/op
ok  	command-line-arguments	1.388s
kalbasit@wks-113 ~/code/src/github.com/icza/gowut/gwu [master] ± % go test -v -bench=. -run=XXX /tmp/a_test.go
PASS
BenchmarkEqual-8	2000000000	         0.33 ns/op
BenchmarkLen-8  	2000000000	         0.34 ns/op
ok  	command-line-arguments	1.413s
kalbasit@wks-113 ~/code/src/github.com/icza/gowut/gwu [master] ± % go test -v -bench=. -run=XXX /tmp/a_test.go
PASS
BenchmarkEqual-8	2000000000	         0.32 ns/op
BenchmarkLen-8  	2000000000	         0.33 ns/op
ok  	command-line-arguments	1.368s
```

I guess it doesn't matter, they perform the same. Go with whichever feel right (style-wize).